File: problems/LeetCode/DP/minPathSum.py
def minPathSum(self, grid: List[List[int]]) -> int:
		ROWS, COLS = len(grid), len(grid[0])

		# [[inf, inf, inf, inf],[inf, inf, inf, inf]] ,,,
		ans = [[float("inf")] * (COLS + 1) for r in range(ROWS + 1)]
		ans[ROWS][COLS - 1] = 0

		for row in range(ROWS -1, -1, -1):
				for col in range(COLS -1, -1, -1):
						ans[row][col] = min(ans[row + 1][col], ans[row][col + 1]) + grid[row][col]

		return ans[0][0]


def minPathSum(self, grid: List[List[int]]) -> int:
		ROWS, COLS = len(grid), len(grid[0])

		dp = [float("inf")] * (COLS + 1)
		dp[COLS - 1] = 0

		for row in range(ROWS - 1, -1, -1):
				for col in range(COLS - 1, -1, -1):
						dp[col] = min(dp[col + 1], dp[col]) + grid[row][col]
				
		return dp[0]


# from top to bottom
def minPathSum(self, grid: List[List[int]]) -> int:

		@cache
		def dfs(r: int, c: int):
				if r == 0 and c == 0:
						return grid[r][c]
				
				if r < 0 or c < 0:
						return inf
				
				return min(dfs(r - 1, c), dfs(r, c - 1)) + grid[r][c]

		return dfs(len(grid) - 1, len(grid[0]) - 1)

File: problems/LeetCode/top100/minPathSum.ts
function minPathSum(grid: number[][]): number {
  const rows = grid.length;
  const cols = grid[0].length;

  const dp = Array.from({ length: rows }, () => Array(cols).fill(0));

  dp[0][0] = grid[0][0];

  // first row
  for (let c = 1; c < cols; c++) {
    dp[0][c] = dp[0][c - 1] + grid[0][c];
  }

  // first col
  for (let r = 1; r < rows; r++) {
    dp[r][0] = dp[r - 1][0] + grid[r][0];
  }

  for (let r = 1; r < rows; r++) {
    for (let c = 1; c < cols; c++) {
      dp[r][c] = grid[r][c] + Math.min(dp[r - 1][c], dp[r][c - 1]);
    }
  }

  return dp[rows - 1][cols - 1];
}

// Time: O(m*n)
// Space: O(m*n)

function minPathSum2(grid: number[][]): number {
  const rows = grid.length;
  const cols = grid[0].length;

  const dp = new Array(cols).fill(0);

  dp[0] = grid[0][0];

  // first row
  for (let c = 1; c < cols; c++) {
    dp[c] = dp[c - 1] + grid[0][c];
  }

  for (let r = 1; r < rows; r++) {
    dp[0] = dp[0] + grid[r][0];
    for (let c = 1; c < cols; c++) {
      dp[c] = grid[r][c] + Math.min(dp[c], dp[c - 1]);
    }
  }

  return dp[cols - 1];
}
2;
// Time: O(m*n)
// Space: O(n)

function minPathSum(grid: number[][]): number {
  const rows = grid.length;
  const cols = grid[0].length;

  const cache = Array.from({ length: rows }, () => Array(cols).fill(-1));

  const dfs = (r: number, c: number) => {
    if (r >= rows || c >= cols) return Infinity;

    if (r === rows - 1 && c === cols - 1) return grid[rows - 1][cols - 1];

    if (cache[r][c] !== -1) return cache[r][c];

    const right = dfs(r, c + 1);
    const bottom = dfs(r + 1, c);

    cache[r][c] = grid[r][c] + Math.min(right, bottom);

    return cache[r][c];
  };

  return dfs(0, 0);
}

// Time: O(m*n)
// Space: O(m*n)
